### ISSUE TITLE
C++ class Counter refactoring

### ISSUE THREAD

**Issue by mislavn** [sysrepo/sysrepo#431]
Hi guys,

I am going to refactor the Counter class so I wanted to know if you @jktjkt , @rastislavszabo, @lukasmacko and @milanlenco have any input on this.

I intend to use templates for Counter class and have a few questions on this subject.

Should the C data structures be accessible to C++ classes only via the Counter class?

Should the Counter class be used everywhere or only where multiple classes share a common C sysrepo data structure?

Regards,
Mislav

**Comment by mislavn** [comment]
I tried to refactor the class Counter to use class templates for constructor parameters.

But then the problem was in passing the the shared_ptr of the Counter class from one class to another. For an example the Data class would need to have the following variables:

std::shared_ptr<Counter<sr_val_t*>>
std::shared_ptr<Counter<sr_val_t*,size_t>>
std::shared_ptr<Counter<sr_val_t**,size_t*>>
std::shared_ptr<Counter<sr_tree_t*>>
std::shared_ptr<Counter<sr_tree_t*,size_t>>
std::shared_ptr<Counter<sr_tree_t**,size_t*>>

This maybe could be fixed with class inheritance, I am not sure.

So for now I simply renamed the Counter class to Deleter (suggestion from @milanlenco ) and add a variable S_Deleter to Data class. This was done in the pull request.

https://github.com/sysrepo/sysrepo/pull/445

**Comment by jktjkt** [comment]
TL;DR: I don't think you can vastly improve this by some refactoring.

Sorry for a late reply; I'll write down some description on what this is all about so that it's easier to find for newcomers in future. Maybe something like that can be moved to in-file source comments? My goal is to ensure that my understandings matches yours; I'm not trying to educate you :).

The ``Counter`` is necessary because of the design choice of using ``std::shared_ptr`` throughout the API. The ``shared_ptr`` doesn't support custom reference counting, i.e., it is not an *intrusive pointer*. This makes it harder to use it for our use case where we might hand out several distinct objects where each of them refer to just a part one underlying structure -- a ``Tree`` hands out ``Val`` instances, and the data can only be freed once all ``Val``s and the original ``Tree`` are gone. That's the job of the ``Counter`` where the actual freeing is not done by our "value classes", but by the auxiliary helper, ``Counter``, whose lifetime is managed for us by the ``shared_ptr`` and which we can control to call a proper cleanup function.

As you identified above, you have to employ some form of dynamic dispatch because a ``Val`` does not necessarily know how to free the resources. This difference must not be reflected in the external-facing ``Val``'s data type, so there must be some type erasure going on. (I.e., there's no ``Val<CreatedFromTree>`` as opposed to ``Val<BoringCreatedDirectly>`` because the user code doesn't really care.)

One method of doing this type erasure is via the "older" method of a ``switch(_t)`` in the destructor. And as you suggest, another valid method which achieves the same goal is through virtual methods where each value-holding class holds a pointer to a specific ``Deleter`` which is an abstract class, and you pass an appropriate ``Deleter<T>``. Each ``Deleter<T>`` inherits from that abstract class, and its destructor frees the memory.

I think that both of these methods are similar -- either you code the distinction by a manual ``switch``, or you use the language's features and use virtual methods to achieve the same goal.

**Comment by mislavn** [comment]
> TL;DR: I don't think you can vastly improve this by some refactoring.

After I tried to refactor this I agree with you on this. So for now I will not make any further attempts at refactoring.

**Comment by rastislavszabo** [comment]
> So for now I will not make any further attempts at refactoring.

Ok, I'm closing this issue.
